### PR TITLE
Implement fix for noUnnecessaryElseRule

### PR DESCRIPTION
**Pull request by eile**

### CODE REVIEW

File: rules/noUnnecessaryElseRule.ts [hunk 1]
```diff
@@ -1,22 +1,54 @@
-import * as ts from 'typescript';
+// tslint
 import * as Lint from 'tslint';
-import * as utils from 'tsutils';
-import { AbstractIfStatementWalker } from '../src/walker';
+
+// tsrules.tslint-consistent-codestyle.src
```
Comment (ajafff) [inline rules/noUnnecessaryElseRule.ts]: please revert these changes

File: rules/noUnnecessaryElseRule.ts [hunk 2]
```diff
@@ -1,22 +1,54 @@
...
-    protected _checkIfStatement(node: ts.IfStatement) {
-        if (node.elseStatement !== undefined &&
-            !isElseIf(node) &&
-            utils.endsControlFlow(node.thenStatement))
-            this.addFailureAtNode(node.getChildAt(5 /*else*/, this.sourceFile), FAIL_MESSAGE);
+class IfWalker extends Lint.AbstractWalker<void> {
+  public walk(sourceFile: ts.SourceFile) {
+    const cb = (node: ts.Node): void => {
+      if (node.kind === ts.SyntaxKind.IfStatement && !this._checkIfStatement(<ts.IfStatement>node))
+        return;
```
Comment (ajafff) [inline rules/noUnnecessaryElseRule.ts]: returning here will not find errors inside `node.condition` and `node.thenStatement`
if the fixed range is smaller we could still inspect all children of this node

File: rules/noUnnecessaryElseRule.ts [hunk 3]
```diff
@@ -1,22 +1,54 @@
...
+
+    return ts.forEachChild(sourceFile, cb);
+  }
+
+  /** Return true to traverse children, false otherwise */
+  private _checkIfStatement(node: ts.IfStatement): boolean {
+    // remove else
+    const elseStatement = node.elseStatement;
+    if (isElseStatement(elseStatement) && !isElseIf(node) && utils.endsControlFlow(node.thenStatement)) {
+      const fixElse = Lint.Replacement.replaceFromTo(node.thenStatement.getEnd(), elseStatement.getStart(), '');
```
Comment (ajafff) [inline rules/noUnnecessaryElseRule.ts]: consider just removing the `else` keyword. this will avoid overlapping fixes

File: rules/noUnnecessaryElseRule.ts [hunk 4]
```diff
@@ -1,22 +1,54 @@
...
+      this.addFailureAtNode(node.getChildAt(5 /*else*/, this.sourceFile), FAIL_MESSAGE, fixElse);
+
+      // remove else scope if safe: we keep multi-statement blocks since the may override local variables when unwrapped
+      if (isElseStatement(elseStatement)  && utils.isBlock(elseStatement) && elseStatement.statements.length === 1) {
```
Comment (ajafff) [inline rules/noUnnecessaryElseRule.ts]: is `isElseStatement` really necessary? we already know `elseStatement` is defined

File: rules/noUnnecessaryElseRule.ts [hunk 5]
```diff
@@ -1,22 +1,54 @@
...
+
+  /** Return true to traverse children, false otherwise */
+  private _checkIfStatement(node: ts.IfStatement): boolean {
+    // remove else
+    const elseStatement = node.elseStatement;
+    if (isElseStatement(elseStatement) && !isElseIf(node) && utils.endsControlFlow(node.thenStatement)) {
+      const fixElse = Lint.Replacement.replaceFromTo(node.thenStatement.getEnd(), elseStatement.getStart(), '');
+      this.addFailureAtNode(node.getChildAt(5 /*else*/, this.sourceFile), FAIL_MESSAGE, fixElse);
+
+      // remove else scope if safe: we keep multi-statement blocks since the may override local variables when unwrapped
```
Comment (ajafff) [inline rules/noUnnecessaryElseRule.ts]: technically that's not correct. consider the following (I know it doesn't make much sense):

```ts
let Foo = 1;
function canNotBeFixed() {
  if (condition) {
    return Foo;
  } else {
    class Foo {} // fixing this would cause a runtime error in the then-branch
  }
}
```

The else-block can be unwrapped if it doesn't contain any block-scoped declarations, regardless of the number of statements

File: rules/noUnnecessaryElseRule.ts [hunk 6]
```diff
@@ -1,22 +1,54 @@
...
+      if (isElseStatement(elseStatement)  && utils.isBlock(elseStatement) && elseStatement.statements.length === 1) {
+        const fixBlock = Lint.Replacement.replaceNode(elseStatement, elseStatement.statements[0].getText());
+        this.addFailureAtNode(node.getChildAt(5 /*else*/, this.sourceFile), FAIL_MESSAGE_BLOCK, fixBlock);
```
Comment (ajafff) [inline rules/noUnnecessaryElseRule.ts]: consider avoiding duplicate error messages

File: rules/noUnnecessaryElseRule.ts [hunk 7]
```diff
@@ -1,22 +1,54 @@
...
+  private _checkIfStatement(node: ts.IfStatement): boolean {
+    // remove else
+    const elseStatement = node.elseStatement;
+    if (isElseStatement(elseStatement) && !isElseIf(node) && utils.endsControlFlow(node.thenStatement)) {
+      const fixElse = Lint.Replacement.replaceFromTo(node.thenStatement.getEnd(), elseStatement.getStart(), '');
+      this.addFailureAtNode(node.getChildAt(5 /*else*/, this.sourceFile), FAIL_MESSAGE, fixElse);
+
+      // remove else scope if safe: we keep multi-statement blocks since the may override local variables when unwrapped
+      if (isElseStatement(elseStatement)  && utils.isBlock(elseStatement) && elseStatement.statements.length === 1) {
+        const fixBlock = Lint.Replacement.replaceNode(elseStatement, elseStatement.statements[0].getText());
```
Comment (ajafff) [inline rules/noUnnecessaryElseRule.ts]: consider just removing `else {` and `}`

File: rules/noUnnecessaryElseRule.ts (hunk 2 above)
Comment (eile) [reply rules/noUnnecessaryElseRule.ts]: Right, but this fixer may need multiple iterations in any case - if you have an ```else if``` cascade with returns in each block, ```endsControlFlow``` will only detect the first instance.

File: rules/noUnnecessaryElseRule.ts (hunk 3 above)
Comment (eile) [reply rules/noUnnecessaryElseRule.ts]: I don't understand what you mean here - this just removes the else already!?

File: rules/noUnnecessaryElseRule.ts (hunk 4 above)
Comment (eile) [reply rules/noUnnecessaryElseRule.ts]: It's to type it to ts.Statement (```node is ts.Statement``` return value)

File: rules/noUnnecessaryElseRule.ts (hunk 6 above)
Comment (eile) [reply rules/noUnnecessaryElseRule.ts]: done

File: rules/noUnnecessaryElseRule.ts (hunk 7 above)
Comment (eile) [reply rules/noUnnecessaryElseRule.ts]: done

File: rules/noUnnecessaryElseRule.ts (hunk 1 above)
Comment (eile) [reply rules/noUnnecessaryElseRule.ts]: done

File: rules/noUnnecessaryElseRule.ts (hunk 2 above)
Comment (ajafff) [reply rules/noUnnecessaryElseRule.ts]: You're right about the need for multiple iterations.
However, I think it's rather surprising if more failures keep popping up after fixing one. Instead I'd like to see all failures at once.

Also note that the following could be fixed in the same iteration if we didn't return early here:

```ts
if (a) {
  if (b) {
    return b;
  } else {
    return b;
  }
} else {
  return c;
}
```

File: rules/noUnnecessaryElseRule.ts [hunk 8]
```diff
@@ -1,22 +1,56 @@
...
+    };
+
+    return ts.forEachChild(sourceFile, cb);
+  }
+
+  /** Return true to traverse children, false otherwise */
+  private _checkIfStatement(node: ts.IfStatement): boolean {
+    const elseStatement = node.elseStatement;
+
+    if (isElseStatement(elseStatement) && !isElseIf(node) && utils.endsControlFlow(node.thenStatement)) {
```
Comment (ajafff) [inline rules/noUnnecessaryElseRule.ts]: you could replace `isElseStatement(elseStatement)` with `elseStatement !== undefined`

File: rules/noUnnecessaryElseRule.ts [hunk 9]
```diff
@@ -1,22 +1,56 @@
...
+    return true;
+  }
+}
+
+function isElseStatement(node: ts.Statement | undefined): node is ts.Statement {
+  return node !== undefined;
+}
+
+function hasLocals(node: ts.Statement): boolean {
+  return !!node.getText().match(/[ \t]+(class|var|let|const)/);
```
Comment (ajafff) [inline rules/noUnnecessaryElseRule.ts]: First things first: `var` is not block scoped, so it can be ignored here.

This sort of text matching is very inaccurate. The declarations don't need to be preceded by a whitespace character, the regex would match text in strings or comments and also match declarations nested in a block or function.

Since you know `node` will always be a `ts.Block`, you could simply iterate through `node.statements` and look for top level `ts.ClassDeclaratation`, `ts.VariableStatement` (where `utils.isBlockScopedVariableDeclarationList(statement.declarationList)`), `ts.EnumDeclaration`, `ts.InterfaceDeclaration` or `ts.TypeAliasDeclaration`

File: rules/noUnnecessaryElseRule.ts [hunk 10]
```diff
@@ -1,22 +1,56 @@
...
+      // remove else scope if safe: keep blocks where local variables change scope when unwrapped
+      if (isElseStatement(elseStatement) && utils.isBlock(elseStatement) && !hasLocals(elseStatement)) {
+        const replacement = elseStatement.getText().replace(elseBraceRegex, '$1').replace(elseRegex, '$1');
+        const fixBlock = Lint.Replacement.replaceNode(elseStatement, replacement);
+        this.addFailureAtNode(node.getChildAt(5 /*else*/, this.sourceFile), FAIL_MESSAGE_BLOCK, fixBlock);
+        return true; // remove the block scope buggers with offsets in children, don't process them now.
+      }
+
+      // remove else
+      const fixElse = Lint.Replacement.replaceFromTo(node.thenStatement.getEnd(), elseStatement.getStart(), '');
```
Comment (ajafff) [inline rules/noUnnecessaryElseRule.ts]: nit: prefer `Lint.Replacement.deleteFromTo`

Comment (ajafff) [inline rules/noUnnecessaryElseRule.ts]: Consider just removing the `else` keyword and leave all the whitespace unchanged.

```ts
if (a)
  a()
else
  b()

// should be fixed to
if (a)
  a()
  b()

// instead of producing invalid code
if (a)
  a()b()
```

File: rules/noUnnecessaryElseRule.ts [hunk 11]
```diff
@@ -1,22 +1,56 @@
...
 import * as Lint from 'tslint';
 import * as utils from 'tsutils';
-import { AbstractIfStatementWalker } from '../src/walker';
+import * as ts from 'typescript';
 import { isElseIf } from '../src/utils';
 
 const FAIL_MESSAGE = `unnecessary else`;
+const FAIL_MESSAGE_BLOCK = `unnecessary else block`;
+
+const elseBraceRegex = /^[ \t]*else[ \t]\{(.*)\}[ \t]*$/;
```
Comment (ajafff) [inline rules/noUnnecessaryElseRule.ts]: consider not using regular expressions. you already have an AST, so you don't need to work with the raw source text. This regex for example doesn't handle comments or line breaks

File: rules/noUnnecessaryElseRule.ts [hunk 12]
```diff
@@ -1,22 +1,56 @@
...
+  }
+
+  /** Return true to traverse children, false otherwise */
+  private _checkIfStatement(node: ts.IfStatement): boolean {
+    const elseStatement = node.elseStatement;
+
+    if (isElseStatement(elseStatement) && !isElseIf(node) && utils.endsControlFlow(node.thenStatement)) {
+      // remove else scope if safe: keep blocks where local variables change scope when unwrapped
+      if (isElseStatement(elseStatement) && utils.isBlock(elseStatement) && !hasLocals(elseStatement)) {
+        const replacement = elseStatement.getText().replace(elseBraceRegex, '$1').replace(elseRegex, '$1');
```
Comment (ajafff) [inline rules/noUnnecessaryElseRule.ts]: I commented on that earlier, so here's a more detailed explanation:
The only reason you need to abort checking child nodes is because you replace the whole statement (fixes for nested nodes are then inside this already fixed range) and TSLint really sucks at handling overlapping fixes.
Actually you only want to remove `else {` and `}`. Here's how I would approach that:

```ts
const elseKeyword = node.getChildAt(5 /*else*/, this.sourceFile);
const fixBlock = [
  Lint.Replacement.deleteFromTo(elseKeyword.end - 4, elseStatement.statements.pos), // removes `else {`
  Lint.Replacement.deleteText(elseStatement.end - 1, 1), // deletes `}`
]
```

**Comment by eile** [comment]
Just saw your comments, reading now..

**Comment by eile** [comment]
Ah, much better now. Thanks for the explanations.

File: test/rules/no-unnecessary-else/default/test.ts.fix [hunk 13]
```diff
@@ -0,0 +1,361 @@
+if (condition) {
+    return;
+}
+
+if (condition) {
+    return;
+}{
+    const foo = "bar";
```
[the remark this answers is not in the document]
Comment (eile) [reply test/rules/no-unnecessary-else/default/test.ts.fix]: all three done

**Comment by ajafff** [comment]
And it's published as `v1.15.0`.
I also reused the logic for error reporting and fixing for `no-else-after-return`.
